src/mbgl/renderer/source_state.cpp

```diff
@@ -0,0 +1,157 @@
+#include <mbgl/renderer/source_state.hpp>
+#include <mbgl/style/conversion_impl.hpp>
+#include <mbgl/renderer/render_tile.hpp>
+#include <mbgl/util/logging.hpp>
+
+namespace mbgl {
+
+using namespace mbgl::style::conversion;
+
+void SourceFeatureState::updateState(const optional<std::string>& sourceLayerID, const std::string& featureID, const Convertible& newState) {
+
+    if (!isObject(newState)) {
+        Log::Error(Event::General, "Feature state must be an object");
+        return;
+    }
+
+    std::string sourceLayer = sourceLayerID ? *sourceLayerID : "";
+
+    std::string stateKey;
+    Value stateValue;
+
+    const std::function<optional<Error> (const std::string&, const Convertible&)> convertFn = [&] (const std::string& k, const Convertible& v) -> optional<Error> {
+        optional<Value> value = toValue(v);
+        if (value) {
+            stateValue = *value;
+        } else {
+            if (isArray(v)) {
+                std::vector<Value> array;
+                size_t length = arrayLength(v);
+                for (size_t i = 0; i < length; i++) {
+                    optional<Value> arrayVal = toValue(arrayMember(v, i));
+                    if (arrayVal) {
+                        array.emplace_back(*arrayVal);
+                    }
+                }
+                std::unordered_map<std::string, Value> result;
+                result[k]= array;
+                stateValue = result;
+                return {};
+
+            } else if (isObject(v)) {
+                eachMember(v, convertFn);
+            }
+        }
+        stateKey = k;
+        PropertyMap stateResult = {{ stateKey, stateValue }};
+        stateChanges[sourceLayer][featureID][stateKey] = stateValue;
+        return {};
+    };
+
+    eachMember(newState, convertFn);
+}
+
+PropertyMap SourceFeatureState::getState(const optional<std::string>& sourceLayerID, const std::string& featureId) const {
+    std::string sourceLayer = sourceLayerID ? *sourceLayerID : "";
+    PropertyMap base, changes;
+    auto states = state.find(sourceLayer);
+    if (states != state.end()) {
+        auto featureStateBase = states->second.find(featureId);
+        if (featureStateBase != states->second.end()) {
+            base = featureStateBase->second;
+        }
+    }
+
+    states = stateChanges.find(sourceLayer);
+    if (states != stateChanges.end()) {
+        auto featureStateChanges = states->second.find(featureId);
+        if (featureStateChanges != states->second.end()) {
+            changes = featureStateChanges->second;
+        }
+    }
+
+    PropertyMap& result = changes;
+    result.insert(base.begin(), base.end());
+    return result;
+}
+
+void SourceFeatureState::coalesceChanges(std::vector<RenderTile>& tiles) {
```

Comment:
This is fairly heavy method, first branch is at least qubic complexity and second is quadratic. If developers would use many states, this would be rather slow. Maybe worth adding performance test to exercise scale-ability of this method.